arch/x86/kvm/pkvm/pgtable.c

```diff
@@ -627,6 +646,60 @@ int pkvm_pgtable_map(struct pkvm_pgtable *pgt, unsigned long vaddr,
 {
 	struct pgt_map_data data = {
 		.prot = prot,
+		.expt_owner_id = NULL,
+		.annotation = 0,
+	};
+	struct pkvm_pgtable_walker walker = {
+		.cb = map_walker,
+		.arg = &data,
+		.walk_flags = PKVM_PGTABLE_WALK_LEAF,
+	};
+
+	if (!VALID_PAGE(phys))
+		return -EINVAL;
+
+	data.phys = ALIGN_DOWN(phys, PAGE_SIZE);
+	return pkvm_pgtable_walk(pgt, vaddr, size, &walker);
+}
+
+/**
+ * pkvm_pgtable_check_and_map() - Installs virtual addr to physical addr
+ *				  mappings in a pkvm pgtable, with validating
+ *				  the owner id first.
+ * @pgt:		The page table to install mappings.
+ * @vaddr:		The virtual address of the installed mapping.
+ * @phys:		The physical address to map.
+ * @size:		The memory size to map.
+ * @prot:		The property bits to create the mapping.
+ * @expt_owner_id:	The expected owner ID stored in the current mapping.
+ *
+ * The checked and mapped range is the minimum PAGE_SIZE-aligned range covering
+ * [@vaddr, @vaddr + @size), and @phys is aligned down to the PAGE_SIZE.
+ *
+ * The mappings for @vaddr to @phys in [@vaddr, @vaddr + @size) are installed
+ * to the page table, with the same @prot. If the mapping already exists at a
+ * higher level, the huge mapping will be split into smaller ones and a new
+ * mapping will be installed at the proper level. If the mapping already exists
+ * at the same level but with different physical address or property bits, it
+ * will be replaced with the new one and TLB will be flushed. Before installing
+ * a mapping for each leaf, check if the owner ID stored in the leaf matches
+ * with @expt_owner_id or not. The mapping will only be installed if they match.
+ * If they don't match, will not install a mapping for this failed leaf and
+ * return the error code (without rollback the mappings for the leaves which are
```

Comment:
Maybe "without removing already installed mappings for the leaves which have passed the check"?